chacha20_poly1305/src/chacha20.rs

```diff
@@ -0,0 +1,460 @@
+/// and there are large performance gains to be had if the function
+/// leverages SIMD instructions on architectures which support them. Because
+/// the algorithm allows for the cipher's state to be operated on in
+/// parallel (each round only touches a quarter of the state), then theoretically
+/// the parallel SIMD instructions should be used. But sometimes the
+/// compiler needs a few hints to ensure it recognizes a "vectorizable" function.
+/// That is the goal of this type, which clearly breaks the state up into four
+/// chunks and exposes functions which align with SIMD lanes.
+///
+/// This type is attempting to be as close as possible to the experimental [`core::simd::u32x4`]
+/// which at this time is feature gated and well beyond the project's MSRV. But ideally
+/// an easy transistion can be made in the future.
+///
+/// A few SIMD relevant design choices:
+///    * Heavy use of inline function calls.
+///    * For-each loops are easy for the compiler to recognize as vectorizable.
+///    * The type is a based on an array instead of tuple since the heterogeneous
+///      nature of tuples can confuse the compiler into thinking it is not vectorizable.
+///    * Memory layout lines up with SIMD size.
+#[repr(C, align(16))]
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct U32x4([u32; 4]);
+
+impl U32x4 {
+    #[inline(always)]
+    fn wrapping_add(self, rhs: Self) -> Self {
+        let mut result = [0u32; 4];
+        (0..4).for_each(|i| {
+            result[i] = self.0[i].wrapping_add(rhs.0[i]);
+        });
+        U32x4(result)
+    }
+
+    #[inline(always)]
+    fn rotate_left(self, n: u32) -> Self {
+        let mut result = [0u32; 4];
+        (0..4).for_each(|i| {
+            result[i] = self.0[i].rotate_left(n);
+        });
+        U32x4(result)
+    }
+
+    #[inline(always)]
+    fn rotate_elements_left<const N: u32>(self) -> Self {
+        let mut result = [0u32; 4];
+        (0..4).for_each(|i| {
+            result[i] = self.0[(i + N as usize) % 4];
+        });
+        U32x4(result)
+    }
+
+    #[inline(always)]
+    fn rotate_elements_right<const N: u32>(self) -> Self {
+        let mut result = [0u32; 4];
+        (0..4).for_each(|i| {
+            result[i] = self.0[(i + 4 - N as usize) % 4];
+        });
+        U32x4(result)
+    }
+
+    #[inline(always)]
+    fn to_le_bytes(self) -> [u8; 16] {
+        let mut bytes = [0u8; 16];
+        bytes[0..4].copy_from_slice(&self.0[0].to_le_bytes());
+        bytes[4..8].copy_from_slice(&self.0[1].to_le_bytes());
+        bytes[8..12].copy_from_slice(&self.0[2].to_le_bytes());
+        bytes[12..16].copy_from_slice(&self.0[3].to_le_bytes());
+        bytes
+    }
+}
+
+impl BitXor for U32x4 {
+    type Output = Self;
+
+    #[inline(always)]
+    fn bitxor(self, rhs: Self) -> Self {
+        let mut result = [0u32; 4];
+        (0..4).for_each(|i| {
+            result[i] = self.0[i] ^ rhs.0[i];
+        });
+        U32x4(result)
+    }
```

Comment:
Could this and `wrapping_add` call into a generic function? `fn binary_op<R>(self, f: Fn(u32, u32) -> u32) -> Self` invoking `f` on each. Should work fine after inlining in theory. But not critical.

Reply:
My goal with the `U32x4` was just to make it as simple as possible so it tracks the std lib version for hopefully an easy switch in the future. I find the binary_op a little much, but wouldn't hate it. Don't think it has to go in now though?

Reply:
I thought there could be more of these operations but probably not so I guess not worth the trouble.